### PR TITLE
Accept invalidated server certificate

### PR DESCRIPTION
**Pull request by martinyunify**

### CODE REVIEW

**Comment by martinyunify** [comment]
@beorn7  add an option to skip handshake validation.

File: cmd/prom2json/main.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ import (
 func main() {
 	cert := flag.String("cert", "", "certificate file")
 	key := flag.String("key", "", "key file")
+	skipCAcheck := flag.Bool("skipCAcheck", false, "Skip certificate check during tls handshake")
```
Comment (beorn7) [inline cmd/prom2json/main.go]: We don't use camel case in flags. This should be `skip-ca-check`. However, I really want strong wording in there that this is insecure and for testing only. Howe about `accept-invalid-cert` and as the doc string: "Accept any certificate during TLS handshake. Insecure, use only for testing."

I'd call the variable something like `skipServerCertCheck` to make clear this is about the server`s cert, not the cert provided above.

File: prom2json.go [hunk 2]
```diff
@@ -137,6 +138,10 @@ func FetchMetricFamilies(
 		}
 		tlsConfig.BuildNameToCertificate()
 		transport = &http.Transport{TLSClientConfig: tlsConfig}
+	} else if skipCACheck {
```
Comment (beorn7) [inline prom2json.go]: `InsecureSkipVerify` isn't mutually exclusive with client certificates, is it?

In principle, you may want to have client auth with a client cert/key pair, while at the same time want to skip the check of the server cert.

**Comment by martinyunify** [comment]
@beorn7 refine the code as suggested.

File: prom2json.go [hunk 3]
```diff
@@ -140,6 +141,13 @@ func FetchMetricFamilies(
 	} else {
 		transport = &http.Transport{}
 	}
+	if skipServerCertCheck {
+		if transport.TLSClientConfig != nil {
+			transport.TLSClientConfig.InsecureSkipVerify = true
+		} else {
+			transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+		}
+	}
```
Comment (beorn7) [inline prom2json.go]: I think you can avoid this whole section by adding a line `InsecureSkipVerify: skipServerCertCheck,` after line 137, and by changing line 142 to
```go
transport = &http.Transport{
    TLSClientConfig: &tls.Config{InsecureSkipVerify: skipServerCertCheck},
}
```

Does that work?

Comment (martinyunify) [reply prom2json.go]: this would erase the configuration object which is created on line 139

**Comment by martinyunify** [comment]
@beorn7 refine the code as suggested.

**Comment by beorn7** [comment]
Great, thanks.

**Pull request merged**